From 9a56620ef46938166e94a7aa048d5fc3066b5fc2 Mon Sep 17 00:00:00 2001
From: Akshay Gupta <akshay.gupta@amd.com>
Date: Fri, 7 Nov 2025 10:16:21 +0000
Subject: [PATCH 07/12] sbrmi: Extend clean hwmon teardown in RMI driver

Update hwmon callbacks to wait for device removal to complete cleanly,
extending the existing clean-exit handling already used by the IOCTL path
Move device bind/unbind handling into a shared helper function
to improve readability, reduce repetition, and preserve existing behavior.

Reviewed-by: Muralidhara M K <muralidhara.mk@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>
Signed-off-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>
---
 drivers/apml/sbrmi.c | 76 ++++++++++++++++++++++++++++++++++++++++-----------------
 1 file changed, 54 insertions(+), 22 deletions(-)

diff --git a/drivers/apml/sbrmi.c b/drivers/apml/sbrmi.c
index 318f8ae..1a5a215 100644
--- a/drivers/apml/sbrmi.c
+++ b/drivers/apml/sbrmi.c
@@ -54,6 +54,33 @@ enum sbrmi_msg_id {
 	SBRMI_READ_PKG_MAX_PWR_LIMIT,
 };
 
+static int sbrmi_prepare_lock(struct apml_sbrmi_device *rmi_dev)
+{
+	mutex_lock(&rmi_dev->lock);
+	/* Verify device unbind/remove is not invoked */
+	if (atomic_read(&rmi_dev->no_new_trans)) {
+		mutex_unlock(&rmi_dev->lock);
+		return -EBUSY;
+	}
+
+	/*
+	 * Set the in_progress variable to true, to wait for
+	 * completion during unbind/remove of driver
+	 */
+	atomic_set(&rmi_dev->in_progress, 1);
+	return 0;
+}
+
+static void sbrmi_prepare_unlock(struct apml_sbrmi_device *rmi_dev)
+{
+	/* Send complete only if device is unbinded/remove */
+	if (atomic_read(&rmi_dev->no_new_trans))
+		complete(&rmi_dev->misc_fops_done);
+
+	atomic_set(&rmi_dev->in_progress, 0);
+	mutex_unlock(&rmi_dev->lock);
+}
+
 static int sbrmi_get_max_pwr_limit(struct apml_sbrmi_device *rmi_dev)
 {
 	struct apml_message msg = { 0 };
@@ -78,6 +105,12 @@ static int sbrmi_read(struct device *dev, enum hwmon_sensor_types type,
 
 	if (type != hwmon_power)
 		return -EINVAL;
+
+	/*
+	 * If device remove/unbind is called do not allow new transaction
+	 */
+	if (atomic_read(&rmi_dev->no_new_trans))
+		return -EBUSY;
 	/* Configure regmap if not configured yet */
 	if (!rmi_dev->regmap) {
 		ret = configure_regmap(rmi_dev);
@@ -87,7 +120,10 @@ static int sbrmi_read(struct device *dev, enum hwmon_sensor_types type,
 		}
 	}
 
-	mutex_lock(&rmi_dev->lock);
+	ret = sbrmi_prepare_lock(rmi_dev);
+	if (ret)
+		return ret;
+
 	msg.data_in.reg_in[RD_FLAG_INDEX] = 1;
 
 	switch (attr) {
@@ -113,7 +149,7 @@ static int sbrmi_read(struct device *dev, enum hwmon_sensor_types type,
 		/* hwmon power attributes are in microWatt */
 		*val = (long)msg.data_out.mb_out[RD_WR_DATA_INDEX] * 1000;
 
-	mutex_unlock(&rmi_dev->lock);
+	sbrmi_prepare_unlock(rmi_dev);
 	return ret;
 }
 
@@ -126,6 +162,12 @@ static int sbrmi_write(struct device *dev, enum hwmon_sensor_types type,
 
 	if (type != hwmon_power && attr != hwmon_power_cap)
 		return -EINVAL;
+
+	/*
+	 * If device remove/unbind is called do not allow new transaction
+	 */
+	if (atomic_read(&rmi_dev->no_new_trans))
+		return -EBUSY;
 	/* Configure regmap if not configured yet */
 	if (!rmi_dev->regmap) {
 		ret = configure_regmap(rmi_dev);
@@ -146,9 +188,13 @@ static int sbrmi_write(struct device *dev, enum hwmon_sensor_types type,
 	msg.data_in.mb_in[RD_WR_DATA_INDEX] = val;
 	msg.data_in.reg_in[RD_FLAG_INDEX] = 0;
 
-	mutex_lock(&rmi_dev->lock);
+	ret = sbrmi_prepare_lock(rmi_dev);
+	if (ret)
+		return ret;
+
 	ret = rmi_mailbox_xfer(rmi_dev, &msg);
-	mutex_unlock(&rmi_dev->lock);
+
+	sbrmi_prepare_unlock(rmi_dev);
 	return ret;
 }
 
@@ -215,23 +261,14 @@ static long sbrmi_ioctl(struct file *fp, unsigned int cmd, unsigned long arg)
 	 * one time. Take lock across so no two protocol is
 	 * invoked at same time, modifying the register value.
 	 */
-	mutex_lock(&rmi_dev->lock);
-	/* Verify device unbind/remove is not invoked */
-	if (atomic_read(&rmi_dev->no_new_trans)) {
-		mutex_unlock(&rmi_dev->lock);
-		return -EBUSY;
-	}
+	ret = sbrmi_prepare_lock(rmi_dev);
+	if (ret)
+		return ret;
 
 	/* Is this a read/monitor/get request */
 	if (msg.data_in.reg_in[RD_FLAG_INDEX])
 		read = true;
 
-	/*
-	 * Set the in_progress variable to true, to wait for
-	 * completion during unbind/remove of driver
-	 */
-	atomic_set(&rmi_dev->in_progress, 1);
-
 	switch (msg.cmd) {
 	case 0 ... 0x999:
 		/* Mailbox protocol */
@@ -260,12 +297,7 @@ static long sbrmi_ioctl(struct file *fp, unsigned int cmd, unsigned long arg)
 		break;
 	}
 
-	/* Send complete only if device is unbinded/remove */
-	if (atomic_read(&rmi_dev->no_new_trans))
-		complete(&rmi_dev->misc_fops_done);
-
-	atomic_set(&rmi_dev->in_progress, 0);
-	mutex_unlock(&rmi_dev->lock);
+	sbrmi_prepare_unlock(rmi_dev);
 
 	/* Copy results back to user only for get/monitor commands and firmware failures */
 	if ((read && !ret) || ret == -EPROTOTYPE) {
-- 
2.43.0

